libraries/AP_InertialNav/AP_InertialNav.cpp

```diff
@@ -28,17 +28,19 @@ void AP_InertialNav::update(bool high_vibes)
 
     // get the velocity relative to the local earth frame
     Vector3f velNED;
-    if (_ahrs_ekf.get_velocity_NED(velNED)) {
-        // during high vibration events use vertical position change
-        if (high_vibes) {
-            float rate_z;
-            if (_ahrs_ekf.get_vert_pos_rate_D(rate_z)) {
-                velNED.z = rate_z;
-            }
+    const bool velned_ok = _ahrs_ekf.get_velocity_NED(velNED);
+    /*
+      during high vibration events use vertical position change. If
+      get_velocity_NED() fails we use a zero horizontal velocity
+    */
+    if (!velned_ok || high_vibes) {
+        float rate_z;
+        if (_ahrs_ekf.get_vert_pos_rate_D(rate_z)) {
+            velNED.z = rate_z;
         }
-        _velocity_cm = velNED * 100; // convert to cm/s
-        _velocity_cm.z = -_velocity_cm.z; // convert from NED to NEU
     }
+    _velocity_cm = velNED * 100; // convert to cm/s
```

Comment:
This means you could be updating X and Y from a bad ekf estimate. You should only update the Z component.

Reply:
we could zero the xy components

Reply:
I don't think zeroing is a good idea, it will result in agresive flying, it would be very bad if we get intermittent values.

We could decay to zero I suppose.

Ultimately i think we just need to stop using X and Y if they go bad.

Reply:
> This means you could be updating X and Y from a bad ekf estimate. You should only update the Z component.

Nothing is changed in how this handles EKF velocities. The EKF will always cause `velned_ok` to be true.
https://github.com/ArduPilot/ardupilot/blob/b86774789b9c3cb3a79f4baeeea25daab83516e4/libraries/AP_AHRS/AP_AHRS.cpp#L3452-L3458
https://github.com/ArduPilot/ardupilot/blob/b86774789b9c3cb3a79f4baeeea25daab83516e4/libraries/AP_AHRS/AP_AHRS.cpp#L345-L345
https://github.com/ArduPilot/ardupilot/blob/422d7cec6e8450a9d15a3b92485bb894caeb0976/libraries/AP_AHRS/AP_AHRS.cpp#L1417-L1427

This current behavior, both in master and this PR, is that the x and y components will be stuck at whatever the last good update was to the `AP_AHRS::state.velocity_NED` variable, and I think that's mostly fine. Position hold hovers are going to do bad things. We can make those modes do less-bad things, but that's another PR.

Reply:
Sure, but this relies on the AHRS behaving in a certain way. We might change that in the future and then hit this untested code path. For example we might stop filling in the vector if returning false which would be seemingly fine because no-one should be using a invalid estimate, then we hit this odd case. 

Better to have a function which is oblivious the same without having to look into outside libs. 

Something like:

```
    // get the velocity relative to the local earth frame
    Vector3f velNED;
    const bool velned_ok = _ahrs_ekf.get_velocity_NED(velNED);
    if (velned_ok) {
        _velocity_cm = velNED * 100.0; // convert to cm/s
        _velocity_cm.z = -_velocity_cm.z; // convert from NED to NEU
    }
    float rate_z;
    if ((!velned_ok || high_vibes) && _ahrs_ekf.get_vert_pos_rate_D(rate_z)) {
        _velocity_cm.z = rate_z * -100.0;
    }